### PR TITLE
Allow cache though enabled repositories to fetch image from upstream

### PR DESCRIPTION
**Pull request by mfuhrmeisterDM**
## what
Add a principal list (`principals_pull_though_access`) which are allowed to use specific repositories as pull through cache (import images from upstream). This holds for repositories where one of the strings in `prefixes_pull_through_repositories` is a prefix of the repository name.

## why
We are using `ecr-public` pull through cache and we want also new images to be downloaded automatically to the cache. Allowed principals for respective repos can use it with the newly introduced variables.

### CODE REVIEW

**Comment by hans-d** [comment]
@mfuhrmeisterDM, thanks ! 
Debatable if this can be considered "read-only" access.

Perhaps better to add a variable for additional rights / or set the default rights, in that case also allowing someone to add the `ecr:TagResource` that might be needed for some pull-through setups.

**Comment by mfuhrmeisterDM** [comment]
> @mfuhrmeisterDM, thanks ! Debatable if this can be considered "read-only" access.
Yes I know, and I'm also more on the side that this not read-only.
 
> Perhaps better to add a variable for additional rights / or set the default rights, in that case also allowing someone to add the `ecr:TagResource` that might be needed for some pull-through setups.

yes good idea, i will prepare something, but I can take until tomorrow.

**Comment by mfuhrmeisterDM** [comment]
May be I will add another list with prefixes which limits it to specific repositories.

**Comment by mfuhrmeisterDM** [comment]
@hans-d I think this is my solution, can you please have another look and run the workflow tests and/or add a review?

**Comment by mfuhrmeisterDM** [comment]
@hans-d any updates regarding this pr?

File: main.tf [hunk 1]
```diff
@@ -292,11 +322,12 @@ data "aws_iam_policy_document" "organization_push_access" {
 }
 
 data "aws_iam_policy_document" "resource" {
-  count = module.this.enabled ? 1 : 0
+  for_each = toset(local.ecr_need_policy && module.this.enabled ? local.image_names : [])
```
[the remark this answers is not in the document]
Comment (mfuhrmeisterDM) [reply main.tf:325]: Right, before the same policy document was used for each image name (ecr-repository). Now some images get extended policy documents. All images with a prefix in the `var.prefixes_pull_through_repositories` will get an extra section where `PullThroughAccess` is allowed for the stated principals in `var.principals_pull_though_access`.

**Comment by mfuhrmeisterDM** [comment]
> Some unfortunate dance re the readme (see checks output). Perhaps you can help with the check re the policy output as well.

readme is fixed

**Review by hans-d — APPROVED** [review]

**Comment by okgolove** [comment]
@mfuhrmeisterDM hello! Could you provide an example how to configure pull through cache using the module? I can't catch what should be put to `prefixes_pull_through_repositories` variable.

**Comment by mfuhrmeisterDM** [comment]
> @mfuhrmeisterDM hello! Could you provide an example how to configure pull through cache using the module? I can't catch what should be put to `prefixes_pull_through_repositories` variable.

```
prefixes_pull_through_repositories:
  - ecr-public
image_names:
  - ecr-public/bitnami/keycloak
  - ecr-public/docker/library/busybox
  - my-company/my-image
```
This is a short example. You have configured a pull through cache rule with `aws_ecr_pull_through_cache_rule` [1]. The prefix used there must be also added to `prefixes_pull_through_repositories`, in this example it is `ecr-public` then all images under `image_names` beginning with one of the prefixes get an extra policy attached that they are allowed to download from the source into the proxy cache [2].

[1] https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecr_pull_through_cache_rule
[2] https://github.com/cloudposse/terraform-aws-ecr/pull/117/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR154-R155

**Pull request merged**